src/dbutils/mutation.rs

```diff
@@ -0,0 +1,359 @@
+    parent: &'m mut Tx,
+    phantom: PhantomData<&'tx i32>,
+    simple_buffer: HashMap<TableName, SimpleBucket>,
+    dupsort_buffer: HashMap<TableName, DupSortBucket>,
+    sequence_increment: HashMap<TableName, u64>,
+}
+
+impl<'tx: 'm, 'm, Tx: MutableTransaction<'tx>> Mutation<'tx, 'm, Tx> {
+    fn new(parent: &'m mut Tx) -> Self {
+        Mutation {
+            parent,
+            phantom: PhantomData,
+            simple_buffer: Default::default(),
+            dupsort_buffer: Default::default(),
+            sequence_increment: Default::default(),
+        }
+    }
+
+    async fn get<T>(&'m self, table: &T, key: &[u8]) -> Result<Option<Bytes<'m>>>
+    where
+        T: Table,
+    {
+        let table_name = table.db_name();
+
+        if let Some(bucket) = self.simple_buffer.get(&table_name) {
+            Ok(match bucket.get(key) {
+                Some(entry) => entry.as_ref().map(|value| value.as_slice().into()),
+                None => self.parent.get(table, key).await?,
+            })
+        } else if let Some(_bucket) = self.dupsort_buffer.get(&table_name) {
+            todo!()
+        } else {
+            Err(MutationError::TableNotFound.into())
+        }
+    }
+
+    async fn read_sequence<T>(&self, table: &T) -> Result<u64>
+    where
+        T: Table,
+    {
+        let parent_value = self.parent.read_sequence(table).await?;
+        let increment = self.sequence_increment.get(&table.db_name()).unwrap_or(&0);
+
+        Ok(parent_value + increment)
+    }
+
+    fn get_simple_buffer(&mut self, name: &TableName) -> &mut SimpleBucket {
+        if !self.simple_buffer.contains_key(name) {
+            self.simple_buffer.insert(name.clone(), Default::default());
+        }
+        self.simple_buffer.get_mut(name).unwrap()
+    }
+
+    fn get_dupsort_buffer(&mut self, name: &TableName) -> &mut DupSortBucket {
+        if !self.dupsort_buffer.contains_key(name) {
+            self.dupsort_buffer.insert(name.clone(), Default::default());
+        }
+        self.dupsort_buffer.get_mut(name).unwrap()
+    }
+
+    fn get_dupsort_changes(&mut self, name: &TableName, k: &[u8]) -> &mut DupSortChanges {
+        let buffer = self.get_dupsort_buffer(name);
+        if !buffer.contains_key(k) {
+            buffer.insert(k.to_vec(), Default::default());
+        }
+        buffer.get_mut(k).unwrap()
+    }
+
+    async fn set<T>(&mut self, table: &T, k: &[u8], v: &[u8]) -> Result<()>
+    where
+        T: Table,
+    {
+        let table_name = table.db_name();
+
+        if is_dupsort(&table_name) {
+            let changes = self.get_dupsort_changes(&table_name, k);
+            changes.insert.insert(v.to_vec());
+            changes.delete.remove(v);
+        } else {
+            self.get_simple_buffer(&table_name)
+                .insert(k.to_vec(), Some(v.to_vec()));
+        }
+
+        Ok(())
+    }
+
+    async fn delete_key<T>(&mut self, table: &T, k: &[u8]) -> Result<()>
```

Comment:
The two delete methods are added since there is no cursor, will we need them both also for dupsort tables or just `delete_pair()`?

Reply:
We will find out only after we start using Mutation in practice 🤷🏻‍♂️